### ISSUE TITLE
Magit is listing ALL versioned files as "New, Staged".

### ISSUE THREAD

**Issue by phil-s** [magit/magit#1713]
I've encountered a very strange bug whereby command line git appears to be completely happy, but Magit (which has been fine until now), is listing every versioned file in the repository under the "Staged" section as a "New" file. 

Status buffer changes are, consequently, super-slow (as the repository contains in excess of 7,000 files).

I'm afraid I don't know what triggered this. How should I debug it? What does Magit do when generating its status data that "git status" does not do?

Magit has no problems with a fresh clone of the repository, reset to the same hash as the one which is exhibiting the problem.

Things I've tried:

* Killing the Magit buffers
* Restarting Emacs
* Upgrading Magit to the latest code from github
* rm .git/index && git reset (original hash)
* There was a file in .git/ which had a filename containing "magit" which I also deleted -- I was pretty hopeful that was the cause, when I saw it -- but that didn't help either.

As best I can tell, Magit regenerates this bad data afresh each time.

Despite having a fresh clone which works, I'd like to restore this working copy to working order, and hopefully understand what actually went wrong here.

**Comment by tarsius** [comment]
Magit uses `git diff-files --cached HEAD` to get the list of staged files.

Try a *hard* reset `git reset --hard <hash>`.

**Comment by phil-s** [comment]
Ok, so looking at `magit-insert-staged-changes' I saw that if `git log -1 HEAD` fails, the diff is performed against an empty tree, which matches the behaviour I was seeing.

Among my untracked files in the broken repository was a file HEAD (which was indeed a copy of .git/HEAD), and git doesn't like that at all.

> fatal: ambiguous argument 'HEAD': both revision and filename

The weird thing is that I *thought* I'd only created that file while I was in the process of resetting the repo to various different points in history, when I originally wondered if I'd managed to break the repository somehow.

I shall presume that I had managed to create this file previously, however, and it was the issue all along.

I don't know whether Magit should handle this or not? Git says:

> Use '--' to separate paths from revisions, like this:
> 'git <command> [<revision>...] -- [<file>...]'

And indeed this works in that situation:

git log -1 HEAD --

But I suspect that would have to be done to a LOT of commands to get around this?

It would have been extremely helpful if the magit process log had included this error. No log is generated for status updates, it seems. This is no doubt because doing so would generate a lot of noise, but I think that *errors* should be logged?

**Comment by tarsius** [comment]
Nice find!

> But I suspect that would have to be done to a LOT of commands to get around this?

Yes, and so far we have added "--" when users reported a particular problem. It might be desirable to do this where ever it makes sense, but getting it right would be a lot of work.

In this particular case I am not doing anything about it. On `next` things blow up somewhere else when a file named `HEAD` exists, and I am no longer anything but the most fatal bugs on `master`.

> It would have been extremely helpful if the magit process log had included this error.

Indeed but that isn't easy. There basically are two ways in which Magit calls Git; (1) for side-effects, in that case output goes into the process log buffer (2) to consume Git's output, in which case neither stdout nor stderr is logged.

Logging stderr in (2) is more complicated and expensive than one would think. It is not possible to send stderr and stdout to two different buffers. Since we consume stdout we have to separate stderr from it. The only way of doing that is to send it to a file and then read it from there.

That's expensive but on `next` I have implemented this a while ago, at least partially.

Unless the exit code is non-zero the stderr is not inserted into the process buffer, so we avoid some work. But creating the temporary file and deleting it (twice, I don't want to talk about it) cannot be avoided.

`magit-git-insert` "consumes" the output by inserting it at point. It is used whenever the output from Git is inserted and then "washed" in place, e.g. when showing diffs and most logs. It might be used more than once when creating a buffer but not as many times as the following functions.

`magit-git-lines` and `magit-git-string` on the other hand are called several dozen time per refresh. Creating a temporary file for each invocation would be a problem, performance wise. And it is quite difficult to implement this properly (I just tried and failed) without introducing bugs. These two functions are used *while reporting an error*, it is very likely that calling them again would again result in the same error and therefor infinite recursion. And they are not only used directly but also by other low-level functions used during error reporting, and it's not so simple to figure out which these functions are. And once we have that list, should we implement two variants of each one of them?

In short, this is surprisingly difficult and I have not yet found a satisfying solution.

So I am closing this as "not the cause just a symptom" :-)

**Comment by phil-s** [comment]
Thanks for looking into that in detail, and sorry to hear that it's so complicated.

I had one thought about the stdout/stderr issue, which was that if we currently don't capture *either* of them, and it's errors that are of the most interest, can we redirect stdout to /dev/null, and capture (to a buffer) *only* stderr? That would seem better than creating temporary files all over the show, and might retain good performance while providing us the most important information?

(I don't suppose that helps with the latter concerns, mind you, but it seemed worth querying regardless.)

**Comment by tarsius** [comment]
I just pushed something to `next`.

There is a new option `magit-git-debug`. When it is nil, which is the default, then behave as before, otherwise do the expensive thing. To avoid infinite recursion let-bind it to nil around dangerous bits.

That's a fairly simple approach, but quite ugly and not completely safe I am afraid. Let's hope it's going to save more problems than new problems it causes.

> if we currently don't capture either of [stdout/stderr], and it's errors that are of the most interest, can we redirect stdout to /dev/null, and capture (to a buffer) only stderr?

When we are not interested we can just use `(nil t)` as "buffer" (see `start-process`) and don't even have to fiddle with `/dev/null`. But that's not a very common case. Also now that I have implemented the above it's no longer advisable to add a special case here (even if it were a common case).

**Comment by phil-s** [comment]
That's great, thank you. I'll be sure to have a play with that sometime soon, and let you know if I encounter any issues.

My only other thought for the moment is that potentially we could distinguish between the instances of (2) (i.e. consuming the output) when we expect a zero exit status in *all* situations (assuming such cases do exist), and instances where we are acting conditionally on the exit status.

If we expect a zero exit status, then a non-zero status could warrant a log message indicating that an unexpected failure has occurred, with a prompt to try repeating the action with `magit-git-debug` enabled.

When a non-zero status is valid, I can't think of a useful way to inform the user that they have debug options available, but I figure informing the user in some cases is better than doing so in none?

**Comment by phil-s** [comment]
And tangentially, `.git/hooks/pre-commit.sample` uses `git rev-parse --verify HEAD` to test for an empty tree, which would seem like a tiny optimisation over using the `log` command for that purpose. (This command also doesn't mind whether a `HEAD` file is present, but you've indicated that's not an isolated issue, so that benefit might be irrelevant at this point.)

In context:
```sh
if git rev-parse --verify HEAD >/dev/null 2>&1
then
	against=HEAD
else
	# Initial commit: diff against an empty tree object
	against=4b825dc642cb6eb9a060e54bf8d69288fbee4904
fi
```

**Comment by tarsius** [comment]
Specifying a revision at all is unnecessary in this case, and not doing it makes all the issues go away. That's what we do on `next`.

In other places we actually do have to verify a rev and there we do it the way you are proposing ... uh wait; while we generally do (`magit-rev-verify`), in this exact case (`magit-no-commit-p`) we don't, let's fix that.
